### PR TITLE
[wip] Add mixed precision

### PR DESCRIPTION
**Pull request by Villux**
Just a draft PR. Based for our future conversations. It would be good if you guys could check this soon and point out if there are some problem areas. Especially `#TODO` parts should be checked. I will add scaling tomorrow and then we might be able to try this already.

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -230,54 +233,61 @@ def train(args, loader, generator, discriminator, g_optim, d_optim, g_ema, devic
...
 
         if g_regularize:
             path_batch_size = max(1, args.batch // args.path_batch_shrink)
             noise = mixing_noise(path_batch_size, args.latent, args.mixing, device)
-            fake_img, latents = generator(noise, return_latents=True)
+            with autocast():
+                fake_img, latents = generator(noise, return_latents=True)
 
+            # TODO scale
+            # https://pytorch.org/docs/stable/notes/amp_examples.html#gradient-penalty
```
Comment (LendelTheGreat) [inline train.py:278]: This is not gradient scaling. I think we don't need to change anything for the g_path_regularize() function.

Comment (Villux) [reply train.py:278]: The values could be so small that they will flush to zero. Not sure if that causes problem in this case. Just want to be sure that the plausible underflow event is not an issue for us.

https://pytorch.org/docs/stable/amp.html#gradient-scaling
>If the forward pass for a particular op has float16 inputs, the backward pass for that op will produce float16 gradients. Gradient values with small magnitudes may not be representable in float16. These values will flush to zero (“underflow”), so the update for the corresponding parameters will be lost.

**Comment by Villux** [comment]
After these changes training is >x2 slower and regularization steps take over a minute :D We need to spent more time to make this work efficiently.  @LendelTheGreat you can continue this while I'm away if you have time.
